core/storage/wal.rs

```diff
@@ -1108,4 +1364,557 @@ impl WalFileShared {
+            res1.num_checkpointed_frames,
+            res1.num_wal_frames
+        );
+
+        // Release reader
+        {
+            let pager = conn2.pager.borrow_mut();
+            let wal2 = pager.wal.borrow_mut();
+            wal2.end_read_tx().unwrap();
+        }
+
+        // Second passive checkpoint should finish
+        let pager = conn1.pager.borrow();
+        let mut wal = pager.wal.borrow_mut();
+        let res2 = run_checkpoint_until_done(&mut *wal, &pager, CheckpointMode::Passive);
+        assert_eq!(
+            res2.num_checkpointed_frames, res2.num_wal_frames,
+            "Second checkpoint completes remaining frames"
+        );
+    }
+
+    #[test]
+    fn test_wal_restart_blocks_readers() {
+        let (db, _) = get_database();
+        let conn1 = db.connect().unwrap();
+        let conn2 = db.connect().unwrap();
+
+        // Start a read transaction
+        conn2
+            .pager
+            .borrow_mut()
+            .wal
+            .borrow_mut()
+            .begin_read_tx()
+            .unwrap();
+
+        // checkpoint should succeed here because the wal is fully checkpointed (empty)
+        // so the reader is using readmark0 to read directly from the db file.
+        let p = conn1.pager.borrow();
+        let mut w = p.wal.borrow_mut();
+        loop {
+            match w.checkpoint(&p, Rc::new(RefCell::new(0)), CheckpointMode::Restart) {
+                Ok(IOResult::IO) => {
+                    conn1.run_once().unwrap();
+                }
+                Ok(IOResult::Done(result)) => {
+                    assert_eq!(
+                        result.num_checkpointed_frames, 0,
+                        "Should not checkpoint any frames (empty)"
+                    );
+                    break;
+                }
+                Err(e) => {
+                    panic!("Checkpoint failed: {e}");
+                }
+            }
+        }
+        drop(w);
+        conn2.pager.borrow_mut().end_read_tx().unwrap();
+
+        conn1
+            .execute("create table test(id integer primary key, value text)")
+            .unwrap();
+        for i in 0..10 {
+            conn1
+                .execute(format!("insert into test(value) values ('value{i}')"))
+                .unwrap();
+        }
+        // now that we have some frames to checkpoint, try again
+        conn2.pager.borrow_mut().begin_read_tx().unwrap();
+        let p = conn1.pager.borrow();
+        let mut w = p.wal.borrow_mut();
+        loop {
+            match w.checkpoint(&p, Rc::new(RefCell::new(0)), CheckpointMode::Restart) {
+                Ok(IOResult::IO) => {
+                    conn1.run_once().unwrap();
+                }
+                Ok(IOResult::Done(_)) => {
+                    panic!("Checkpoint should not have succeeded");
+                }
+                Err(e) => {
+                    assert!(matches!(e, LimboError::Busy), "should block readers");
```

Comment:
This is not true, right? It's more like `"should return Busy if any readers are using the WAL"`? Same re: the test name.